File: app/domain.py
"""Pure, deterministic domain rules. No DB, HTTP, or model calls."""
from collections import Counter
from copy import deepcopy

GRADES = ["Junior", "Middle", "Senior", "Lead"]
REPEATABLE_EVENTS = {"EV_036"}


def role_key(role, grade):
    return f"{role}|{grade}"


def employee_history(data, employee_id):
    return sorted(
        [r for r in data["history"].values() if r["employee_id"] == employee_id],
        key=lambda r: (r["date"], r.get("completed_at", ""), r["record_id"]),
    )


def apply_gains(levels, event):
    result = dict(levels)
    for gain in event["develops_skills"]:
        sid = gain["skill_id"]
        old = result.get(sid, 0)
        result[sid] = old + max(0, min(gain["gain"], gain["max_level"] - old, 5 - old))
    return result


def effective_skills(employee, data, as_of):
    levels = {sid: employee["skills"].get(sid, 0) for sid in data["skills"]}
    seen = set()
    for row in employee_history(data, employee["employee_id"]):
        if row["record_id"] in seen:
            continue
        seen.add(row["record_id"])
        after_review = row["date"] > employee["last_review_date"]
        # A locally completed action on the snapshot date follows the snapshot assessment.
        same_day_action = row.get("source") == "application" and row["date"] == employee["last_review_date"]
        if row["status"] == "completed" and row["date"] <= as_of and (after_review or same_day_action):
            levels = apply_gains(levels, data["events"][row["event_id"]])
    return levels


def target_profile(employee, data):
    goal = employee.get("career_goal")
    source = "explicit"
    if goal is None:
        index = GRADES.index(employee["grade"])
        if index == len(GRADES) - 1:
            return {"goal": None, "source": "not_set", "requirements": {}, "critical_skills": []}
        goal = {"target_role": employee["role"], "target_grade": GRADES[index + 1]}
        source = "suggested_next_grade"
    profile = data["role_profiles"].get(role_key(goal["target_role"], goal["target_grade"]))
    return {
        "goal": goal, "source": source,
        "requirements": profile["required_skills"] if profile else {},
        "critical_skills": profile["critical_skills"] if profile else [],
    }


def trajectory(employee, data, as_of):
    levels = effective_skills(employee, data, as_of)
    target = target_profile(employee, data)
    gaps = []
    for sid, required in target["requirements"].items():
        gaps.append({
            "skill_id": sid, "name": data["skills"][sid]["name"],
            "current": levels.get(sid, 0), "required": required,
            "gap": max(0, required - levels.get(sid, 0)),
            "critical": sid in target["critical_skills"],
        })
    total = sum(target["requirements"].values())
    covered = sum(min(x["current"], x["required"]) for x in gaps)
    return {
        "effective_skills": levels, "target": target,
        "skills": sorted(gaps, key=lambda x: (not x["critical"], -x["gap"], x["skill_id"])),
        "coverage_pct": round(100 * covered / total, 2) if total else None,
        "critical_gaps": [x for x in gaps if x["critical"] and x["gap"] > 0],
    }


def eligibility(employee, event, levels, history, as_of):
    reasons = []
    if event["mandatory"]:
        reasons.append("mandatory")
    if employee["role"] not in event["target_roles"]:
        reasons.append("role_not_eligible")
    if employee["grade"] not in event["target_grades"]:
        reasons.append("grade_not_eligible")
    if any(levels.get(sid, 0) < required for sid, required in event["prerequisites"].items()):
        reasons.append("prerequisites_not_met")
    completed = [r for r in history if r["event_id"] == event["event_id"] and r["status"] == "completed"]
    if completed and event["event_id"] not in REPEATABLE_EVENTS:
        reasons.append("already_completed")
    ongoing = [r for r in history if r["event_id"] == event["event_id"] and r["status"] == "in_progress"]
    completed_dates = {r.get("session_date", r["date"]) for r in completed}
    sessions = sorted(d for d in event["upcoming_sessions"] if d >= as_of and d not in completed_dates)
    if event["format"] != "self_paced" and not sessions and not ongoing:
        reasons.append("no_available_session")
    return reasons, ongoing, sessions


def candidates(employee, data, as_of):
    progress = trajectory(employee, data, as_of)
    levels = progress["effective_skills"]
    history = employee_history(data, employee["employee_id"])
    requirements = progress["target"]["requirements"]
    result, excluded = [], {}
    for event in data["events"].values():
        reasons, ongoing, sessions = eligibility(employee, event, levels, history, as_of)
        after = apply_gains(levels, event)
        gains = [
            {"skill_id": sid, "before": levels.get(sid, 0), "after": value, "gain": value - levels.get(sid, 0),
             "gap_closed": min(max(0, requirements.get(sid, 0) - levels.get(sid, 0)), value - levels.get(sid, 0)),
             "critical": sid in progress["target"]["critical_skills"]}
            for sid, value in after.items() if value > levels.get(sid, 0)
        ]
        if not progress["target"]["goal"]:
            reasons.append("goal_not_set")
        elif not any(g["gap_closed"] for g in gains):
            reasons.append("no_target_gain")
        if reasons:
            excluded[event["event_id"]] = reasons
            continue
        result.append({
            **deepcopy(event), "expected_gains": gains,
            "action": "continue" if ongoing else "start",
            "activity_record_id": ongoing[0]["record_id"] if ongoing else None,
            "next_session": ongoing[0]["date"] if ongoing else (sessions[0] if sessions else None),
        })
    # Stable input order only. AI, not this ordering, is responsible for recommendations.
    result.sort(key=lambda x: x["event_id"])
    return result, excluded


def profile_view(employee, data, as_of):
    available, excluded = candidates(employee, data, as_of)
    return {
        "employee": deepcopy(employee), "as_of_date": as_of,
        "trajectory": trajectory(employee, data, as_of),
        "history": employee_history(data, employee["employee_id"]),
        "available_steps": available,
        "no_step_reasons": dict(Counter(reason for reasons in excluded.values() for reason in reasons)) if not available else {},
    }


def recommendation_context(employee, data, as_of, revision):
    view = profile_view(employee, data, as_of)
    types, formats = {}, {}
    for row in view["history"]:
        if row["date"] > as_of:
            continue
        event = data["events"][row["event_id"]]
        for group, key in [(types, event["type"]), (formats, event["format"])]:
            group.setdefault(key, Counter())[row["status"]] += 1
    return {
        "schema_version": "1.0", "data_revision": revision, "as_of_date": as_of,
        "employee": {key: employee[key] for key in ("employee_id", "role", "grade", "tenure_months", "work_format", "preferred_language")},
        "trajectory": view["trajectory"], "candidates": view["available_steps"],
        "history_summary": {"by_type": types, "by_format": formats, "total_records": len(view["history"])},
        "no_step_reasons": view["no_step_reasons"],
    }


def hr_overview(data, as_of, date_from=None, date_to=None):
    deficits, no_steps = {}, []
    for employee in data["employees"].values():
        progress = trajectory(employee, data, as_of)
        for skill in progress["skills"]:
            row = deficits.setdefault(skill["skill_id"], {
                "skill_id": skill["skill_id"], "name": skill["name"], "target_population": 0,
                "employees_with_gap": 0, "critical_gaps": 0,
            })
            row["target_population"] += 1
            if skill["gap"]:
                row["employees_with_gap"] += 1
                row["critical_gaps"] += int(skill["critical"])
        available, excluded = candidates(employee, data, as_of)
        if not available:
            no_steps.append({
                "employee_id": employee["employee_id"], "full_name": employee["full_name"],
                "role": employee["role"], "grade": employee["grade"],
                "reasons": dict(Counter(r for reasons in excluded.values() for r in reasons)),
                "goal": progress["target"]["goal"], "coverage_pct": progress["coverage_pct"],
            })
    for row in deficits.values():
        row["gap_pct"] = round(100 * row["employees_with_gap"] / row["target_population"], 2)
    participation = {eid: {"event_id": eid, "title": e["title"], "records": 0, "statuses": {s: 0 for s in ("completed", "in_progress", "dropped", "no_show", "declined", "overdue")}} for eid, e in data["events"].items()}
    for row in data["history"].values():
        if row["date"] > (date_to or as_of) or (date_from and row["date"] < date_from):
            continue
        item = participation[row["event_id"]]
        item["records"] += 1
        item["statuses"][row["status"]] += 1
    for item in participation.values():
        item["completion_pct"] = round(100 * item["statuses"]["completed"] / item["records"], 2) if item["records"] else None
    return {
        "as_of_date": as_of, "employee_count": len(data["employees"]),
        "skill_deficits": sorted(deficits.values(), key=lambda x: (-x["employees_with_gap"], x["skill_id"])),
        "employees_without_step": no_steps,
        "participation": list(participation.values()),
        "period": {"from": date_from, "to": date_to or as_of},
        "definitions": {"gap_pct": "employees_with_gap / target_population × 100", "completion_pct": "completed records / all records in period × 100"},
    }

File: frontend/domain.js
// Read-only previews of published server rules. Never authorizes an API action.
const number=value=>typeof value==='number'&&Number.isFinite(value);
const ordered=rows=>[...rows].sort((a,b)=>String(a.date).localeCompare(String(b.date))||String(a.completed_at||'').localeCompare(String(b.completed_at||''))||String(a.record_id).localeCompare(String(b.record_id)));
const history=profile=>profile.raw_history||profile.history||[];
const catalog=profile=>profile.catalog_events||[];
const names=profile=>profile.skill_names||Object.fromEntries((profile.all_skills||[]).map(s=>[s.skill_id,s.name]));
const requirements=profile=>profile.raw_trajectory?.target?.requirements;
const critical=profile=>profile.raw_trajectory?.target?.critical_skills||(profile.skills||[]).filter(s=>s.critical).map(s=>s.skill_id);
const level=(levels,id)=>levels[id]??0;
function validLevels(levels){return levels&&typeof levels==='object'&&Object.values(levels).every(v=>number(v)&&v>=0&&v<=5);}
function hasTarget(profile){const req=requirements(profile);return !!profile.goal&&req&&Object.keys(req).length>0&&Object.values(req).every(v=>number(v)&&v>=0&&v<=5);}
function metrics(profile,levels){
  const req=requirements(profile),total=Object.values(req||{}).reduce((a,b)=>a+b,0);
  return {coverage:total?Math.round(10000*Object.entries(req).reduce((sum,[id,value])=>sum+Math.min(level(levels,id),value),0)/total)/100:null,critical:critical(profile).filter(id=>level(levels,id)<(req?.[id]??0)).length};
}
function capped(levels,event){
  const after={...levels};
  if(!Array.isArray(event.develops_skills))return null;
  for(const item of event.develops_skills){
    if(!item.skill_id||!number(item.gain)||!number(item.max_level)||item.gain<0||item.max_level<0||item.max_level>5)return null;
    const old=level(after,item.skill_id);after[item.skill_id]=old+Math.max(0,Math.min(item.gain,item.max_level-old,5-old));
  }
  return after;
}
function eligibility(profile,event,levels,rows=history(profile),{target=true}={}){
  const reasons=[],employee=profile.employee||{},asof=profile.as_of_date;
  if(event.mandatory)reasons.push('mandatory');
  if(!Array.isArray(event.target_roles)||!Array.isArray(event.target_grades)||!event.prerequisites||!asof||!employee.role||!employee.grade)reasons.push('incomplete_data');
  else {
    if(!event.target_roles.includes(employee.role))reasons.push('role_not_eligible');
    if(!event.target_grades.includes(employee.grade))reasons.push('grade_not_eligible');
    if(Object.entries(event.prerequisites).some(([id,value])=>level(levels,id)<value))reasons.push('prerequisites_not_met');
  }
  const completed=rows.filter(r=>r.event_id===event.event_id&&r.status==='completed');
  if(completed.length&&event.event_id!=='EV_036')reasons.push('already_completed');
  const ongoing=rows.some(r=>r.event_id===event.event_id&&r.status==='in_progress');
  const completedDates=new Set(completed.map(r=>r.session_date||r.date));
  if(event.format!=='self_paced'&&!ongoing&&!(event.upcoming_sessions||[]).some(d=>d>=asof&&!completedDates.has(d)))reasons.push('no_available_session');
  if(target){
    if(!hasTarget(profile))reasons.push('goal_not_set');
    else {const after=capped(levels,event),req=requirements(profile);if(!after)reasons.push('incomplete_data');else if(!Object.keys(req).some(id=>level(levels,id)<req[id]&&level(after,id)>level(levels,id)))reasons.push('no_target_gain');}
  }
  return [...new Set(reasons)];
}
function enrich(profile,event){
  const skillNames=names(profile);
  return {...event,description:event.description||'',develops_skills:(event.develops_skills||[]).map(g=>({...g,name:skillNames[g.skill_id]||g.skill_id})),prerequisite_details:Object.entries(event.prerequisites||{}).map(([id,required])=>({skill_id:id,name:skillNames[id]||id,required,current:level(profile.effective_skills||{},id)})),gains:(event.expected_gains||event.gains||[]).map(g=>({...g,name:skillNames[g.skill_id]||g.name||g.skill_id}))};
}
export function eventById(profile,id){
  const entry=catalog(profile).find(e=>e.event_id===id),available=(profile.available_steps||[]).find(e=>e.event_id===id),rec=profile.recommendations?.items?.find(e=>e.event_id===id);
  if(!entry&&!available&&!rec)return null;
  const event=enrich(profile,{...entry,...rec,...available,event_id:id});
  const ongoing=history(profile).find(r=>r.event_id===id&&r.status==='in_progress');
  // Only genuine server entries/records can provide an action. Forecast unlocks cannot.
  if(available){event.action=available.action;event.record_id=available.record_id||available.activity_record_id;if(!Object.hasOwn(available,'expected_gains'))delete event.expected_gains;}
  else if(ongoing&&!event.mandatory){event.action='continue';event.record_id=ongoing.record_id;event.can_complete=validLevels(profile.effective_skills)&&!eligibility(profile,event,profile.effective_skills,history(profile),{target:false}).length&&(event.format==='self_paced'||ongoing.date<=profile.as_of_date);}
  else {delete event.action;delete event.record_id;event.can_complete=false;}
  const reasons=validLevels(profile.effective_skills)?eligibility(profile,event,profile.effective_skills):['incomplete_data'];
  return {...event,eligible:!!available,reasons:available?[]:reasons};
}
const unavailable=reason=>({available:false,reason,reasons:[reason],coverage_before:null,coverage_after:null,critical_before:null,critical_after:null,changes:[],unlocks:[]});
function delta(profile,before,after){const skillNames=names(profile);return Object.keys(after).filter(id=>level(after,id)!==level(before,id)).map(id=>({skill_id:id,name:skillNames[id]||id,before:level(before,id),after:level(after,id)}));}
function newlyAvailable(profile,before,after,rows){
  const previous=new Set(catalog(profile).filter(e=>!eligibility(profile,e,before,history(profile)).length).map(e=>e.event_id));
  return catalog(profile).filter(e=>!previous.has(e.event_id)&&!eligibility(profile,e,after,rows).length).map(e=>({...enrich(profile,e),eligible:false,simulated:true,action:undefined,record_id:undefined,can_complete:false}));
}
export function forecast(profile,input){
  const event=typeof input==='string'?eventById(profile,input):eventById(profile,input?.event_id);
  if(!event||!validLevels(profile.effective_skills)||!hasTarget(profile))return unavailable('incomplete_data');
  const ongoing=history(profile).some(r=>r.event_id===event.event_id&&r.status==='in_progress');
  const reasons=eligibility(profile,event,profile.effective_skills,history(profile),{target:!ongoing});
  if(reasons.length)return {...unavailable(reasons[0]),reasons};
  // An entry that would be suitable locally still cannot become an actionable server recommendation.
  if(!event.eligible&&!ongoing)return unavailable('not_available');
  const before={...profile.effective_skills};let after;
  if(Array.isArray(event.expected_gains)){
    after={...before};
    for(const gain of event.expected_gains){if(gain.before!==level(before,gain.skill_id)||!number(gain.after)||gain.after<gain.before||gain.after>5)return unavailable('stale_forecast');after[gain.skill_id]=gain.after;}
  }else after=capped(before,event);
  if(!after)return unavailable('incomplete_data');
  const initial=metrics(profile,before),final=metrics(profile,after);
  const rows=history(profile).filter(r=>!(r.event_id===event.event_id&&r.status==='in_progress')).concat({event_id:event.event_id,status:'completed',date:event.session_date||profile.as_of_date});
  return {available:true,reason:null,reasons:[],coverage_before:initial.coverage,coverage_after:final.coverage,critical_before:initial.critical,critical_after:final.critical,changes:delta(profile,before,after),unlocks:newlyAvailable(profile,before,after,rows)};
}
export function filterEvents(profile,{tab='suitable',search='',format='',duration='',skills=[]}={}){
  let ids=tab==='ongoing'?history(profile).filter(r=>r.status==='in_progress').map(r=>r.event_id):tab==='all'?catalog(profile).filter(e=>!e.mandatory).map(e=>e.event_id):(profile.available_steps||[]).map(e=>e.event_id);
  const query=search.toLocaleLowerCase().trim(),limit=Number(String(duration).replace(/[^\d.]/g,''));
  return [...new Set(ids)].map(id=>eventById(profile,id)).filter(Boolean).filter(e=>!e.mandatory&&(!query||[e.title,e.description,...e.develops_skills.map(g=>g.name)].join(' ').toLocaleLowerCase().includes(query))&&(!format||e.format===format||e.type===format)&&(!limit||e.duration_hours<=limit)&&(!skills.length||skills.some(id=>e.develops_skills.some(g=>g.skill_id===id))));
}
function replay(profile){
  const employee=profile.employee||{};
  if(!employee.last_review_date||!profile.as_of_date||!validLevels(employee.skills))return {available:false,reason:'assessment_unavailable'};
  let levels={...employee.skills};const rows=[],seen=new Set();
  for(const record of ordered(history(profile))){
    if(seen.has(record.record_id))continue;seen.add(record.record_id);
    if(record.status!=='completed'||record.date>profile.as_of_date||!(record.date>employee.last_review_date||(record.date===employee.last_review_date&&record.source==='application')))continue;
    const event=catalog(profile).find(e=>e.event_id===record.event_id),before={...levels},after=event?capped(levels,event):null;
    if(!after)return {available:false,reason:'event_history_unavailable',rows};
    rows.push({record,event,before,after});levels=after;
  }
  return {available:true,base:{...employee.skills},rows,levels};
}
export function skillTimeline(profile,skillId){
  const result=replay(profile),employee=profile.employee||{};
  if(!result.available)return [{kind:'unavailable',date:null,title:'История расчёта недоступна',before:null,after:null,gain:null,source:'unknown',explanation:result.reason}];
  const base=level(result.base,skillId),items=[{kind:'assessment',date:employee.last_review_date,title:'Базовая оценка',before:null,after:base,gain:null,source:'assessment'}];
  for(const {record,event,before,after} of result.rows){if(level(after,skillId)===level(before,skillId))continue;items.push({kind:'activity',date:record.date,title:event.title,record_id:record.record_id,event_id:event.event_id,before:level(before,skillId),after:level(after,skillId),gain:level(after,skillId)-level(before,skillId),source:record.source||'import'});}
  items.push({kind:'current',date:profile.as_of_date,title:'Уровень на дату среза',before:null,after:level(result.levels,skillId),gain:null,source:'calculated'});return items;
}
export function historyResult(profile,recordId){
  const exact=profile.completion_results?.[recordId];if(exact)return {...exact,available:true,source:'completion'};
  const result=replay(profile);if(!result.available)return unavailable(result.reason);
  const step=result.rows.find(r=>r.record.record_id===recordId);
  if(!step)return unavailable('included_in_assessment_or_not_completed');
  const before=hasTarget(profile)?metrics(profile,step.before):{coverage:null,critical:null},after=hasTarget(profile)?metrics(profile,step.after):{coverage:null,critical:null};
  return {available:true,source:'reconstructed',event_id:step.event.event_id,record_id:recordId,title:step.event.title,changes:delta(profile,step.before,step.after),coverage_before:before.coverage,coverage_after:after.coverage,critical_before:before.critical,critical_after:after.critical,unlocks:[],explanation:'Расчёт по текущему каталогу и текущей цели. Историческая цель и прежняя версия каталога не сохранены.'};
}
